src/cc/frontends/clang/kbuild_helper.cc

```diff
@@ -107,4 +109,78 @@ int KBuildHelper::get_flags(const char *uname_machine, vector<string> *cflags) {
   return 0;
 }
 
+static inline int file_exists(const char *f)
+{
+  struct stat buffer;
+  return (stat(f, &buffer) == 0);
+}
+
+static inline int proc_kheaders_exists(void)
+{
+  return file_exists(PROC_KHEADERS_PATH);
+}
+
+static inline int extract_kheaders(std::string &dirpath)
+{
+  char tar_cmd[256];
+  int ret;
+  bool module = false;
+
+  if (!proc_kheaders_exists()) {
+    ret = system("modprobe kheaders");
+    if (ret)
+      return ret;
+    if (!proc_kheaders_exists())
+      return -1;
+    module = true;
+  }
+
+  dirpath = "/tmp/proc-kheaders-XXXXXX";
+  if (mkdtemp((char *)dirpath.data()) == NULL)
+    return -1;
+
+  snprintf(tar_cmd, 256, "tar -xf %s -C %s", PROC_KHEADERS_PATH,
+      dirpath.c_str());
+
+  ret = system(tar_cmd);
+  if (ret) {
+    system(("rm -rf " + dirpath).c_str());
+    return ret;
+  }
+
+  // Mark extraction as completed
+  std::ofstream output(dirpath + "/extracted");
+
+  if (module) {
+    ret = system("rmmod kheaders");
+    if (ret)
+      return ret;
+  }
+  return ret;
+}
+
+int get_proc_kheaders(std::string &dirpath)
+{
+  // Check if kheaders already exist and reuse it
+  DIR *dir;
+  struct dirent *ent;
+  if ((dir = opendir("/tmp/")) != NULL) {
```

Comment:
that doesn't work for systems with multiple kernels and corresponding kernel headers installed.
The host may reboot into a different kernel and wrong headers will be picked up.
I suggest to use a fingerprint of /proc/kheaders.tar.xz and extract it into fixed location like:
/tmp/proc-kheaders-sha1(kheaders.txz)/

Reply:
I can do that, but my thought was the /tmp directory is cleaned up after a reboot anyway and if a new kernel is installed, then a reboot is needed. The clearing may not be the default though on all systems...

I can do the md5 approach you suggested.. Thanks.

Reply:
bcc is already using sha1 for such fingerprinting purpose. See bpf_prog_compute_tag.
Another alternative is to use uname -r plus some other property of kheaders file as part of the dir.

Reply:
When trying to use the code in `bpf_prog_compute_tag`, I get:
```
sha1 socket not available Address family not supported by protocol
```
I am worried this is due to missing kernel support and will break on Android.  We do have md5sum binary available on the filesystem, so I will just call that binary. md5sum is pretty universal in my experience.

Reply:
and how subsequent bcc run will know the directory name? every time it tries to compile bpf prog it will do md5sum of /proc/kheader.tar.xz ? I think we need a cheaper way to construct dirname.

Reply:
Subsequent run does a strstr search of kheaders-<hash> in /tmp. The file being hashed is in memory, md5sum is very fast in my tests - just 14-15ms ; which is a negligible cost.

Reply:
> file being hashed is in memory

don't quite follow.
when /tmp/kheaders-TBD is present bcc shouldn't be loading kmod, reading /proc/kheader and doing md5sum on it. modprobe is not a cheap operation.
May be just uname -r is enough.
Or uname -r plus crc of /proc/config.gz That should give stable dir without modprobe of kheaders

Reply:
Ok. I agree that's better. I will rework patches accordingly and resend.

Reply:
Instead of relying on marker file ("/extracted") and doing reading /tmp (which can be pretty big, my devmachine has 8500 files right now, and it's not the max), why not atomically rename "kheaders-<sha>-<rndsuffix>" into a "canonical" "kheaders-<sha>" folder. Then the presence of this well-defined directory would be a marker of success. And then even in case of race, we'll end up still having just one directory with good contents. Do you think it will work?

Reply:
I don't see a benefit of such approach , tbh, my current approach works so I'm happy with it. You could post a follow up patch if you want with your approach to discuss your idea, once my patch is merged.

Reply:
As I see it, your current approach has two downsides:
1. It requires iterating over all contents of /tmp directory, which might be slow, depending on the state of /tmp on user's host. It's also unnecessarily pollutes kernel's dentry/inode cache. And it's just a bunch of code that can be eliminated.
2. I know you are aware, as you mentioned this possibility above, but it still has races and can leave multiple directories with uncompressed headers. By doing atomic rename, there is still race, but it won't leave extra directories lying around.
Both seem to be pretty clear disadvantages, which could be easily avoided.

Reply:
Ok I think I know what you mean. But it still has the disadvantage that during a race, there may be 2 rename attempts and the failure of one of them will have to be special-handled. I revised these patches 4-5 times and I am currently approaching a burn out :) I prefer any future such extensions as you propose, can be worked on top of this patch, since I don't see the drawbacks you mentioned as anything major. The race is quite rare, and in the such an event, the only cost is an extra extraction. So I don't see a big benefit. I don't mind seeing if I can work on that as a follow-up patch (unless you can), but I don't see the benefit of doing it in this patch other than being for bike-shedding.

@4ast are you Ok with this patch now as-is ?

Reply:
@joelagnel I'm confused why mktemp and readdir are still in the patch.
I thought we agreed on stable name:
/tmp/kheaders-`uname -r`-`crc of /proc/config.gz`

Reply:
@4ast because if 2 instances of a tool are running, then they may try to extract at the same time to same directory and can cause issues. The parallel extractions need to happen in their own tmp directories. The only way around that is to use a lock file or something.. does that make sense?

Also I prefer to just stick to /tmp/kheaders-uname and not to take CRC of /proc/config.gz because that may break if IKCONFIG is disabled in the .config. Let me know if you have any objection to anything I said.

Reply:
to solve two parallel untars how about let it extract into temp directory and then rename into a dir with stable name? If rename fails we just rm that untared dir that lost the race.
if we assume that /tmp is cleaned on every reboot then /tmp/kheaders-uname is fine.

Reply:
Yes, ok, I will do that. So the plan is mktemp into a directory, and then rename it to a stable one. If the rename fails, then just delete the directory. I will do that then and update the PR, thanks.